# FVC endmember calibration from Earth Engine

This diagnostic notebook recalculates the statistical NDVI endmembers used for fractional vegetation cover (FVC) **from the satellite products and repository code**.

It does **not** depend on previous CSV or Excel diagnostic files.

The notebook remains outside the production pipeline. Its required production output is:

`config/fvc_endmembers.json`

The production pipeline should fail explicitly if this calibration file is absent or invalid.

## Fixed calibration method

For each optical source (`HLS` and `S2`) independently:

1. Use the MODIS 8-day periods from the study interval.
2. Build the source-specific optical medoid for every period using the current repository modules.
3. Define optical coverage from the common valid mask of `Green`, `Red`, and `NIR`.
4. Retain station-period observations with `coverage_pct >= 80`.
5. Calculate NDVI and NDWI from the medoid.
6. Exclude water pixels using `NDWI > 0`.
7. Within each valid station-period:
   - low candidate = NDVI P05
   - high candidate = NDVI P95
8. Across all valid 2021–2023 station-period observations:
   - statistical low-NDVI endmember = P05 of all station-period P05 values
   - statistical high-NDVI endmember = P95 of all station-period P95 values

The endmembers are global in time and source-specific.


In [21]:
from __future__ import annotations

import json
import sys
import time
from datetime import date, timedelta
from pathlib import Path

import ee
import pandas as pd


## 1. Repository setup and Earth Engine initialization

The notebook searches upward for the repository root, so it can be executed with a clean kernel from `notebooks/diagnostics/`.


In [22]:
def find_repo_root(start: Path | None = None) -> Path:
    current = (start or Path.cwd()).resolve()

    for candidate in [current, *current.parents]:
        if (candidate / "pyproject.toml").exists():
            return candidate

    raise FileNotFoundError(
        "Repository root not found. Expected to find pyproject.toml."
    )


REPO_ROOT = find_repo_root()
SRC_PATH = REPO_ROOT / "src"

if str(SRC_PATH) not in sys.path:
    sys.path.insert(0, str(SRC_PATH))

print("Repository:", REPO_ROOT)
print("Source path:", SRC_PATH)


Repository: C:\Users\User\Desktop\cristian\GEE\26-et-downscaling-fundacion
Source path: C:\Users\User\Desktop\cristian\GEE\26-et-downscaling-fundacion\src


In [23]:
EE_PROJECT = "ee-change"

ee.Initialize(project=EE_PROJECT)
ee.Number(1).getInfo()

print("Earth Engine initialized with project:", EE_PROJECT)


Earth Engine initialized with project: ee-change


## 2. Import the current repository processing functions

The calibration deliberately reuses the current optical preprocessing and medoid functions instead of reimplementing sensor QA in the notebook.


In [24]:
from et_downscaling.hls import (
    build_hls_medoid,
    get_hls_collection,
)
from et_downscaling.modis import build_modis_inputs
from et_downscaling.sentinel2 import (
    build_s2_medoid,
    get_sentinel2_collection,
)


## 3. Calibration settings

`outputs/diagnostics/` is used only for ignored diagnostic checkpoints and QA tables.

`config/fvc_endmembers.json` is the versioned calibration contract consumed by the production pipeline.


In [25]:
COVERAGE_THRESHOLD_PCT = 80.0
SOURCE_SCALES_M = {
    "HLS": 30,
    "S2": 20,
}

WITHIN_PERIOD_LOW_PERCENTILE = 5
WITHIN_PERIOD_HIGH_PERCENTILE = 95
GLOBAL_LOW_QUANTILE = 0.05
GLOBAL_HIGH_QUANTILE = 0.95

# Set True to ignore this notebook's own checkpoint and recompute every period.
FORCE_REBUILD = False

DIAGNOSTIC_DIR = REPO_ROOT / "outputs" / "diagnostics"
CHECKPOINT_PATH = (
    DIAGNOSTIC_DIR / "_fvc_endmember_calibration_checkpoint.csv"
)
OBSERVATION_OUTPUT_PATH = (
    DIAGNOSTIC_DIR / "fvc_endmember_calibration_observations.csv"
)
ENDMEMBER_TABLE_PATH = (
    DIAGNOSTIC_DIR / "fvc_global_endmembers.csv"
)
PIPELINE_CONFIG_PATH = (
    REPO_ROOT / "config" / "fvc_endmembers.json"
)

DIAGNOSTIC_DIR.mkdir(parents=True, exist_ok=True)
PIPELINE_CONFIG_PATH.parent.mkdir(parents=True, exist_ok=True)

print("Checkpoint:", CHECKPOINT_PATH)
print("Pipeline calibration:", PIPELINE_CONFIG_PATH)


Checkpoint: C:\Users\User\Desktop\cristian\GEE\26-et-downscaling-fundacion\outputs\diagnostics\_fvc_endmember_calibration_checkpoint.csv
Pipeline calibration: C:\Users\User\Desktop\cristian\GEE\26-et-downscaling-fundacion\config\fvc_endmembers.json


## 4. Build station footprints and optical collections from source data

No previous diagnostic table is loaded here.


In [26]:
modis_inputs = build_modis_inputs()

modis_collection = ee.ImageCollection(modis_inputs["collection"])
station_footprints = ee.FeatureCollection(
    modis_inputs["station_footprints"]
)

s2_collection = get_sentinel2_collection(station_footprints)
hls_collection = get_hls_collection(station_footprints)

station_count = int(station_footprints.size().getInfo())

station_properties = station_footprints.select(
    [
        "station",
        "station_id",
        "longitude",
        "latitude",
        "footprint_area_m2",
    ]
).getInfo()["features"]

stations = [
    feature["properties"]
    for feature in station_properties
]

print("Stations:", station_count)
print([station["station"] for station in stations])


Stations: 5
['Pastos limpios', 'Palma', 'Bananera', 'Manglar', 'Bosque seco']


## 5. Get the MODIS 8-day calibration periods

The optical endmembers are calibrated on the same temporal support used by the ET target.


In [27]:
period_millis = (
    modis_collection
    .aggregate_array("system:time_start")
    .getInfo()
)

period_starts = sorted(
    {
        pd.to_datetime(
            value,
            unit="ms",
            utc=True,
        ).date()
        for value in period_millis
    }
)


def get_period_end(period_start: date) -> date:
    regular_end = period_start + timedelta(days=8)
    next_year_start = date(period_start.year + 1, 1, 1)
    return min(regular_end, next_year_start)


periods = [
    (period_start, get_period_end(period_start))
    for period_start in period_starts
]

print("MODIS periods:", len(periods))
print("First:", periods[0])
print("Last:", periods[-1])


MODIS periods: 138
First: (datetime.date(2021, 1, 1), datetime.date(2021, 1, 9))
Last: (datetime.date(2023, 12, 27), datetime.date(2024, 1, 1))


## 6. Earth Engine helper functions

Coverage is computed before the water mask from the common valid support of `Green`, `Red`, and `NIR`.

The water mask is applied only to the NDVI distribution used for endmember candidates.


In [28]:
def safe_normalized_difference(
    image: ee.Image,
    positive_band: str,
    negative_band: str,
    output_name: str,
    epsilon: float = 1e-6,
) -> ee.Image:
    image = ee.Image(image)

    positive = image.select(positive_band)
    negative = image.select(negative_band)

    numerator = positive.subtract(negative)
    denominator = positive.add(negative)

    return (
        numerator
        .divide(denominator)
        .updateMask(denominator.abs().gt(epsilon))
        .rename(output_name)
        .toFloat()
    )


def build_ndvi_ndwi(medoid: ee.Image) -> ee.Image:
    medoid = ee.Image(medoid)

    ndvi = safe_normalized_difference(
        medoid,
        "NIR",
        "Red",
        "NDVI",
    )

    ndwi = safe_normalized_difference(
        medoid,
        "Green",
        "NIR",
        "NDWI",
    )

    return ndvi.addBands(ndwi)


def get_common_valid_mask(medoid: ee.Image) -> ee.Image:
    return (
        ee.Image(medoid)
        .select(["Green", "Red", "NIR"])
        .mask()
        .reduce(ee.Reducer.min())
        .rename("valid")
    )


def get_info_with_retry(
    ee_object,
    attempts: int = 5,
    base_sleep_seconds: int = 5,
):
    last_error = None

    for attempt in range(1, attempts + 1):
        try:
            return ee_object.getInfo()

        except Exception as error:
            last_error = error

            if attempt == attempts:
                break

            sleep_seconds = base_sleep_seconds * attempt
            print(
                f"Earth Engine request failed "
                f"(attempt {attempt}/{attempts}). "
                f"Retrying in {sleep_seconds}s..."
            )
            time.sleep(sleep_seconds)

    raise last_error


## 7. Calculate one source-period group

For every period, the medoid is built once for the union of the five station footprints. Statistics are then evaluated independently for each footprint.

Low-coverage observations remain in the diagnostic table, but their NDVI endmember candidates are not used.


In [29]:
def build_source_period_rows(
    source: str,
    period_start: date,
    period_end: date,
) -> list[dict]:
    if source == "HLS":
        full_collection = hls_collection
        build_medoid = build_hls_medoid

    elif source == "S2":
        full_collection = s2_collection
        build_medoid = build_s2_medoid

    else:
        raise ValueError(f"Unsupported optical source: {source}")

    scale = SOURCE_SCALES_M[source]
    start_text = period_start.isoformat()
    end_text = period_end.isoformat()

    period_collection = (
        ee.ImageCollection(full_collection)
        .filterDate(start_text, end_text)
        .filterBounds(station_footprints.geometry())
    )

    product_count = int(
        get_info_with_retry(period_collection.size())
    )

    # No optical products: record five explicit zero-coverage observations.
    if product_count == 0:
        rows = []

        for station in stations:
            rows.append(
                {
                    "source": source,
                    "station": station.get("station"),
                    "station_id": station.get("station_id"),
                    "period_start": start_text,
                    "period_end": end_text,
                    "source_scale_m": scale,
                    "products": 0,
                    "coverage_pct": 0.0,
                    "NDVI_p05": None,
                    "NDVI_p95": None,
                    "nonwater_pixel_count": 0,
                }
            )

        return rows

    analysis_geometry = station_footprints.geometry()

    medoid = ee.Image(
        build_medoid(
            period_collection,
            analysis_geometry,
        )
    )

    indices = build_ndvi_ndwi(medoid)

    valid_mask = get_common_valid_mask(medoid)
    ndvi = indices.select("NDVI")
    ndwi = indices.select("NDWI")

    nonwater_ndvi = ndvi.updateMask(
        ndwi.lte(0)
    )

    projection = medoid.select("NIR").projection()

    percentile_reducer = (
        ee.Reducer.percentile(
            [
                WITHIN_PERIOD_LOW_PERCENTILE,
                WITHIN_PERIOD_HIGH_PERCENTILE,
            ]
        )
        .combine(
            reducer2=ee.Reducer.count(),
            sharedInputs=True,
        )
    )

    def add_statistics(feature):
        feature = ee.Feature(feature)
        geometry = feature.geometry()

        coverage_raw = (
            valid_mask
            .unmask(0)
            .reduceRegion(
                reducer=ee.Reducer.mean(),
                geometry=geometry,
                crs=projection,
                scale=scale,
                maxPixels=1e6,
                tileScale=4,
            )
            .get("valid")
        )

        coverage_pct = ee.Number(
            ee.Algorithms.If(
                ee.Algorithms.IsEqual(
                    coverage_raw,
                    None,
                ),
                0,
                ee.Number(coverage_raw).multiply(100),
            )
        )

        raw_ndvi_statistics = ee.Dictionary(
            ee.Algorithms.If(
                coverage_pct.gte(
                    COVERAGE_THRESHOLD_PCT
                ),
                nonwater_ndvi.reduceRegion(
                    reducer=percentile_reducer,
                    geometry=geometry,
                    crs=projection,
                    scale=scale,
                    maxPixels=1e6,
                    tileScale=4,
                ),
                ee.Dictionary({}),
            )
        )

        # Some high-coverage footprints can contain zero non-water pixels.
        # In that case Earth Engine omits percentile keys entirely.
        # Fill missing keys with explicit sentinels so the mapped
        # FeatureCollection can be evaluated without failing.
        ndvi_statistics = ee.Dictionary(
            {
                "p5": -9999,
                "p95": -9999,
                "count": 0,
            }
        ).combine(
            raw_ndvi_statistics,
            overwrite=True,
        )

        return feature.set(
            {
                "source": source,
                "period_start": start_text,
                "period_end": end_text,
                "source_scale_m": scale,
                "products": product_count,
                "coverage_pct": coverage_pct,
                "NDVI_p05": ndvi_statistics.get("p5"),
                "NDVI_p95": ndvi_statistics.get("p95"),
                "nonwater_pixel_count": ndvi_statistics.get(
                    "count"
                ),
            }
        )

    result = ee.FeatureCollection(
        station_footprints.map(add_statistics)
    )

    info = get_info_with_retry(result)

    rows = []

    for feature in info["features"]:
        properties = feature["properties"]

        rows.append(
            {
                "source": properties.get("source"),
                "station": properties.get("station"),
                "station_id": properties.get("station_id"),
                "period_start": properties.get("period_start"),
                "period_end": properties.get("period_end"),
                "source_scale_m": properties.get(
                    "source_scale_m"
                ),
                "products": properties.get("products"),
                "coverage_pct": properties.get(
                    "coverage_pct"
                ),
                "NDVI_p05": properties.get("NDVI_p05"),
                "NDVI_p95": properties.get("NDVI_p95"),
                "nonwater_pixel_count": properties.get(
                    "nonwater_pixel_count"
                ),
            }
        )

    return rows


## 8. Recalculate all HLS and Sentinel-2 station-period observations

This cell is restartable. Its checkpoint is generated by this notebook and lives under `outputs/`, which can remain ignored by Git.

With no checkpoint present, the notebook starts from the satellite products and recalculates everything.


In [30]:
checkpoint_columns = [
    "source",
    "station",
    "station_id",
    "period_start",
    "period_end",
    "source_scale_m",
    "products",
    "coverage_pct",
    "NDVI_p05",
    "NDVI_p95",
    "nonwater_pixel_count",
]

if FORCE_REBUILD and CHECKPOINT_PATH.exists():
    CHECKPOINT_PATH.unlink()

if CHECKPOINT_PATH.exists():
    observations = pd.read_csv(CHECKPOINT_PATH)
    print(
        "Loaded notebook checkpoint:",
        len(observations),
        "rows",
    )
else:
    observations = pd.DataFrame(
        columns=checkpoint_columns
    )
    print("No checkpoint found. Starting from source data.")


def group_is_complete(
    table: pd.DataFrame,
    source: str,
    period_start: str,
) -> bool:
    subset = table.loc[
        (table["source"] == source)
        & (table["period_start"] == period_start)
    ]

    return subset["station_id"].nunique() == station_count


total_groups = len(periods) * len(SOURCE_SCALES_M)
group_index = 0

for source in SOURCE_SCALES_M:
    for period_start, period_end in periods:
        group_index += 1
        start_text = period_start.isoformat()

        if group_is_complete(
            observations,
            source,
            start_text,
        ):
            print(
                f"{group_index}/{total_groups} | "
                f"{source} | {start_text} | cached"
            )
            continue

        print(
            f"{group_index}/{total_groups} | "
            f"{source} | "
            f"{start_text} -> {period_end.isoformat()}"
        )

        rows = build_source_period_rows(
            source,
            period_start,
            period_end,
        )

        # Replace a partial group if a previous run stopped mid-write.
        if not observations.empty:
            observations = observations.loc[
                ~(
                    (observations["source"] == source)
                    & (
                        observations["period_start"]
                        == start_text
                    )
                )
            ].copy()

        observations = pd.concat(
            [
                observations,
                pd.DataFrame(rows),
            ],
            ignore_index=True,
        )

        observations = observations[
            checkpoint_columns
        ].sort_values(
            [
                "source",
                "period_start",
                "station_id",
            ]
        )

        observations.to_csv(
            CHECKPOINT_PATH,
            index=False,
        )

print("Final station-period rows:", len(observations))


Loaded notebook checkpoint: 5 rows
1/276 | HLS | 2021-01-01 | cached
2/276 | HLS | 2021-01-09 -> 2021-01-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


3/276 | HLS | 2021-01-17 -> 2021-01-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


4/276 | HLS | 2021-01-25 -> 2021-02-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


5/276 | HLS | 2021-02-02 -> 2021-02-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


6/276 | HLS | 2021-02-10 -> 2021-02-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


7/276 | HLS | 2021-02-18 -> 2021-02-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


8/276 | HLS | 2021-02-26 -> 2021-03-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


9/276 | HLS | 2021-03-06 -> 2021-03-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


10/276 | HLS | 2021-03-14 -> 2021-03-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


11/276 | HLS | 2021-03-22 -> 2021-03-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


12/276 | HLS | 2021-03-30 -> 2021-04-07


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


13/276 | HLS | 2021-04-07 -> 2021-04-15


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


14/276 | HLS | 2021-04-15 -> 2021-04-23


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


15/276 | HLS | 2021-04-23 -> 2021-05-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


16/276 | HLS | 2021-05-01 -> 2021-05-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


17/276 | HLS | 2021-05-09 -> 2021-05-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


18/276 | HLS | 2021-05-17 -> 2021-05-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


19/276 | HLS | 2021-05-25 -> 2021-06-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


20/276 | HLS | 2021-06-02 -> 2021-06-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


21/276 | HLS | 2021-06-10 -> 2021-06-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


22/276 | HLS | 2021-06-18 -> 2021-06-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


23/276 | HLS | 2021-06-26 -> 2021-07-04


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


24/276 | HLS | 2021-07-04 -> 2021-07-12


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


25/276 | HLS | 2021-07-12 -> 2021-07-20


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


26/276 | HLS | 2021-07-20 -> 2021-07-28


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


27/276 | HLS | 2021-07-28 -> 2021-08-05


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


28/276 | HLS | 2021-08-05 -> 2021-08-13


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


29/276 | HLS | 2021-08-13 -> 2021-08-21


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


30/276 | HLS | 2021-08-21 -> 2021-08-29


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


31/276 | HLS | 2021-08-29 -> 2021-09-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


32/276 | HLS | 2021-09-06 -> 2021-09-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


33/276 | HLS | 2021-09-14 -> 2021-09-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


34/276 | HLS | 2021-09-22 -> 2021-09-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


35/276 | HLS | 2021-09-30 -> 2021-10-08


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


36/276 | HLS | 2021-10-08 -> 2021-10-16


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


37/276 | HLS | 2021-10-16 -> 2021-10-24


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


38/276 | HLS | 2021-10-24 -> 2021-11-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


39/276 | HLS | 2021-11-01 -> 2021-11-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


40/276 | HLS | 2021-11-09 -> 2021-11-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


41/276 | HLS | 2021-11-17 -> 2021-11-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


42/276 | HLS | 2021-11-25 -> 2021-12-03


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


43/276 | HLS | 2021-12-03 -> 2021-12-11


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


44/276 | HLS | 2021-12-11 -> 2021-12-19


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


45/276 | HLS | 2021-12-19 -> 2021-12-27


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


46/276 | HLS | 2021-12-27 -> 2022-01-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


47/276 | HLS | 2022-01-01 -> 2022-01-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


48/276 | HLS | 2022-01-09 -> 2022-01-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


49/276 | HLS | 2022-01-17 -> 2022-01-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


50/276 | HLS | 2022-01-25 -> 2022-02-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


51/276 | HLS | 2022-02-02 -> 2022-02-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


52/276 | HLS | 2022-02-10 -> 2022-02-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


53/276 | HLS | 2022-02-18 -> 2022-02-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


54/276 | HLS | 2022-02-26 -> 2022-03-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


55/276 | HLS | 2022-03-06 -> 2022-03-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


56/276 | HLS | 2022-03-14 -> 2022-03-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


57/276 | HLS | 2022-03-22 -> 2022-03-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


58/276 | HLS | 2022-03-30 -> 2022-04-07


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


59/276 | HLS | 2022-04-07 -> 2022-04-15


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


60/276 | HLS | 2022-04-15 -> 2022-04-23


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


61/276 | HLS | 2022-04-23 -> 2022-05-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


62/276 | HLS | 2022-05-01 -> 2022-05-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


63/276 | HLS | 2022-05-09 -> 2022-05-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


64/276 | HLS | 2022-05-17 -> 2022-05-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


65/276 | HLS | 2022-05-25 -> 2022-06-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


66/276 | HLS | 2022-06-02 -> 2022-06-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


67/276 | HLS | 2022-06-10 -> 2022-06-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


68/276 | HLS | 2022-06-18 -> 2022-06-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


69/276 | HLS | 2022-06-26 -> 2022-07-04


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


70/276 | HLS | 2022-07-04 -> 2022-07-12


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


71/276 | HLS | 2022-07-12 -> 2022-07-20


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


72/276 | HLS | 2022-07-20 -> 2022-07-28


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


73/276 | HLS | 2022-07-28 -> 2022-08-05


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


74/276 | HLS | 2022-08-05 -> 2022-08-13


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


75/276 | HLS | 2022-08-13 -> 2022-08-21


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


76/276 | HLS | 2022-08-21 -> 2022-08-29


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


77/276 | HLS | 2022-08-29 -> 2022-09-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


78/276 | HLS | 2022-09-06 -> 2022-09-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


79/276 | HLS | 2022-09-14 -> 2022-09-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


80/276 | HLS | 2022-09-22 -> 2022-09-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


81/276 | HLS | 2022-09-30 -> 2022-10-08


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


82/276 | HLS | 2022-10-08 -> 2022-10-16


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


83/276 | HLS | 2022-10-16 -> 2022-10-24


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


84/276 | HLS | 2022-10-24 -> 2022-11-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


85/276 | HLS | 2022-11-01 -> 2022-11-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


86/276 | HLS | 2022-11-09 -> 2022-11-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


87/276 | HLS | 2022-11-17 -> 2022-11-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


88/276 | HLS | 2022-11-25 -> 2022-12-03


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


89/276 | HLS | 2022-12-03 -> 2022-12-11


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


90/276 | HLS | 2022-12-11 -> 2022-12-19


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


91/276 | HLS | 2022-12-19 -> 2022-12-27


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


92/276 | HLS | 2022-12-27 -> 2023-01-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


93/276 | HLS | 2023-01-01 -> 2023-01-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


94/276 | HLS | 2023-01-09 -> 2023-01-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


95/276 | HLS | 2023-01-17 -> 2023-01-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


96/276 | HLS | 2023-01-25 -> 2023-02-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


97/276 | HLS | 2023-02-02 -> 2023-02-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


98/276 | HLS | 2023-02-10 -> 2023-02-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


99/276 | HLS | 2023-02-18 -> 2023-02-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


100/276 | HLS | 2023-02-26 -> 2023-03-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


101/276 | HLS | 2023-03-06 -> 2023-03-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


102/276 | HLS | 2023-03-14 -> 2023-03-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


103/276 | HLS | 2023-03-22 -> 2023-03-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


104/276 | HLS | 2023-03-30 -> 2023-04-07


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


105/276 | HLS | 2023-04-07 -> 2023-04-15


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


106/276 | HLS | 2023-04-15 -> 2023-04-23


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


107/276 | HLS | 2023-04-23 -> 2023-05-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


108/276 | HLS | 2023-05-01 -> 2023-05-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


109/276 | HLS | 2023-05-09 -> 2023-05-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


110/276 | HLS | 2023-05-17 -> 2023-05-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


111/276 | HLS | 2023-05-25 -> 2023-06-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


112/276 | HLS | 2023-06-02 -> 2023-06-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


113/276 | HLS | 2023-06-10 -> 2023-06-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


114/276 | HLS | 2023-06-18 -> 2023-06-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


115/276 | HLS | 2023-06-26 -> 2023-07-04


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


116/276 | HLS | 2023-07-04 -> 2023-07-12


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


117/276 | HLS | 2023-07-12 -> 2023-07-20


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


118/276 | HLS | 2023-07-20 -> 2023-07-28


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


119/276 | HLS | 2023-07-28 -> 2023-08-05


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


120/276 | HLS | 2023-08-05 -> 2023-08-13


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


121/276 | HLS | 2023-08-13 -> 2023-08-21


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


122/276 | HLS | 2023-08-21 -> 2023-08-29


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


123/276 | HLS | 2023-08-29 -> 2023-09-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


124/276 | HLS | 2023-09-06 -> 2023-09-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


125/276 | HLS | 2023-09-14 -> 2023-09-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


126/276 | HLS | 2023-09-22 -> 2023-09-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


127/276 | HLS | 2023-09-30 -> 2023-10-08


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


128/276 | HLS | 2023-10-08 -> 2023-10-16


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


129/276 | HLS | 2023-10-16 -> 2023-10-24


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


130/276 | HLS | 2023-10-24 -> 2023-11-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


131/276 | HLS | 2023-11-01 -> 2023-11-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


132/276 | HLS | 2023-11-09 -> 2023-11-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


133/276 | HLS | 2023-11-17 -> 2023-11-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


134/276 | HLS | 2023-11-25 -> 2023-12-03


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


135/276 | HLS | 2023-12-03 -> 2023-12-11


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


136/276 | HLS | 2023-12-11 -> 2023-12-19


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


137/276 | HLS | 2023-12-19 -> 2023-12-27


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


138/276 | HLS | 2023-12-27 -> 2024-01-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


139/276 | S2 | 2021-01-01 -> 2021-01-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


140/276 | S2 | 2021-01-09 -> 2021-01-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


141/276 | S2 | 2021-01-17 -> 2021-01-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


142/276 | S2 | 2021-01-25 -> 2021-02-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


143/276 | S2 | 2021-02-02 -> 2021-02-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


144/276 | S2 | 2021-02-10 -> 2021-02-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


145/276 | S2 | 2021-02-18 -> 2021-02-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


146/276 | S2 | 2021-02-26 -> 2021-03-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


147/276 | S2 | 2021-03-06 -> 2021-03-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


148/276 | S2 | 2021-03-14 -> 2021-03-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


149/276 | S2 | 2021-03-22 -> 2021-03-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


150/276 | S2 | 2021-03-30 -> 2021-04-07


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


151/276 | S2 | 2021-04-07 -> 2021-04-15


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


152/276 | S2 | 2021-04-15 -> 2021-04-23


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


153/276 | S2 | 2021-04-23 -> 2021-05-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


154/276 | S2 | 2021-05-01 -> 2021-05-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


155/276 | S2 | 2021-05-09 -> 2021-05-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


156/276 | S2 | 2021-05-17 -> 2021-05-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


157/276 | S2 | 2021-05-25 -> 2021-06-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


158/276 | S2 | 2021-06-02 -> 2021-06-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


159/276 | S2 | 2021-06-10 -> 2021-06-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


160/276 | S2 | 2021-06-18 -> 2021-06-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


161/276 | S2 | 2021-06-26 -> 2021-07-04


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


162/276 | S2 | 2021-07-04 -> 2021-07-12


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


163/276 | S2 | 2021-07-12 -> 2021-07-20


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


164/276 | S2 | 2021-07-20 -> 2021-07-28


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


165/276 | S2 | 2021-07-28 -> 2021-08-05


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


166/276 | S2 | 2021-08-05 -> 2021-08-13


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


167/276 | S2 | 2021-08-13 -> 2021-08-21


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


168/276 | S2 | 2021-08-21 -> 2021-08-29


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


169/276 | S2 | 2021-08-29 -> 2021-09-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


170/276 | S2 | 2021-09-06 -> 2021-09-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


171/276 | S2 | 2021-09-14 -> 2021-09-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


172/276 | S2 | 2021-09-22 -> 2021-09-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


173/276 | S2 | 2021-09-30 -> 2021-10-08


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


174/276 | S2 | 2021-10-08 -> 2021-10-16


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


175/276 | S2 | 2021-10-16 -> 2021-10-24


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


176/276 | S2 | 2021-10-24 -> 2021-11-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


177/276 | S2 | 2021-11-01 -> 2021-11-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


178/276 | S2 | 2021-11-09 -> 2021-11-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


179/276 | S2 | 2021-11-17 -> 2021-11-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


180/276 | S2 | 2021-11-25 -> 2021-12-03


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


181/276 | S2 | 2021-12-03 -> 2021-12-11


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


182/276 | S2 | 2021-12-11 -> 2021-12-19


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


183/276 | S2 | 2021-12-19 -> 2021-12-27


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


184/276 | S2 | 2021-12-27 -> 2022-01-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


185/276 | S2 | 2022-01-01 -> 2022-01-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


186/276 | S2 | 2022-01-09 -> 2022-01-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


187/276 | S2 | 2022-01-17 -> 2022-01-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


188/276 | S2 | 2022-01-25 -> 2022-02-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


189/276 | S2 | 2022-02-02 -> 2022-02-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


190/276 | S2 | 2022-02-10 -> 2022-02-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


191/276 | S2 | 2022-02-18 -> 2022-02-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


192/276 | S2 | 2022-02-26 -> 2022-03-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


193/276 | S2 | 2022-03-06 -> 2022-03-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


194/276 | S2 | 2022-03-14 -> 2022-03-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


195/276 | S2 | 2022-03-22 -> 2022-03-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


196/276 | S2 | 2022-03-30 -> 2022-04-07


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


197/276 | S2 | 2022-04-07 -> 2022-04-15


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


198/276 | S2 | 2022-04-15 -> 2022-04-23


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


199/276 | S2 | 2022-04-23 -> 2022-05-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


200/276 | S2 | 2022-05-01 -> 2022-05-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


201/276 | S2 | 2022-05-09 -> 2022-05-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


202/276 | S2 | 2022-05-17 -> 2022-05-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


203/276 | S2 | 2022-05-25 -> 2022-06-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


204/276 | S2 | 2022-06-02 -> 2022-06-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


205/276 | S2 | 2022-06-10 -> 2022-06-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


206/276 | S2 | 2022-06-18 -> 2022-06-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


207/276 | S2 | 2022-06-26 -> 2022-07-04


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


208/276 | S2 | 2022-07-04 -> 2022-07-12


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


209/276 | S2 | 2022-07-12 -> 2022-07-20


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


210/276 | S2 | 2022-07-20 -> 2022-07-28


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


211/276 | S2 | 2022-07-28 -> 2022-08-05


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


212/276 | S2 | 2022-08-05 -> 2022-08-13


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


213/276 | S2 | 2022-08-13 -> 2022-08-21


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


214/276 | S2 | 2022-08-21 -> 2022-08-29


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


215/276 | S2 | 2022-08-29 -> 2022-09-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


216/276 | S2 | 2022-09-06 -> 2022-09-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


217/276 | S2 | 2022-09-14 -> 2022-09-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


218/276 | S2 | 2022-09-22 -> 2022-09-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


219/276 | S2 | 2022-09-30 -> 2022-10-08


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


220/276 | S2 | 2022-10-08 -> 2022-10-16


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


221/276 | S2 | 2022-10-16 -> 2022-10-24


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


222/276 | S2 | 2022-10-24 -> 2022-11-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


223/276 | S2 | 2022-11-01 -> 2022-11-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


224/276 | S2 | 2022-11-09 -> 2022-11-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


225/276 | S2 | 2022-11-17 -> 2022-11-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


226/276 | S2 | 2022-11-25 -> 2022-12-03


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


227/276 | S2 | 2022-12-03 -> 2022-12-11


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


228/276 | S2 | 2022-12-11 -> 2022-12-19


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


229/276 | S2 | 2022-12-19 -> 2022-12-27


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


230/276 | S2 | 2022-12-27 -> 2023-01-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


231/276 | S2 | 2023-01-01 -> 2023-01-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


232/276 | S2 | 2023-01-09 -> 2023-01-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


233/276 | S2 | 2023-01-17 -> 2023-01-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


234/276 | S2 | 2023-01-25 -> 2023-02-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


235/276 | S2 | 2023-02-02 -> 2023-02-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


236/276 | S2 | 2023-02-10 -> 2023-02-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


237/276 | S2 | 2023-02-18 -> 2023-02-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


238/276 | S2 | 2023-02-26 -> 2023-03-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


239/276 | S2 | 2023-03-06 -> 2023-03-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


240/276 | S2 | 2023-03-14 -> 2023-03-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


241/276 | S2 | 2023-03-22 -> 2023-03-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


242/276 | S2 | 2023-03-30 -> 2023-04-07


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


243/276 | S2 | 2023-04-07 -> 2023-04-15


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


244/276 | S2 | 2023-04-15 -> 2023-04-23


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


245/276 | S2 | 2023-04-23 -> 2023-05-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


246/276 | S2 | 2023-05-01 -> 2023-05-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


247/276 | S2 | 2023-05-09 -> 2023-05-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


248/276 | S2 | 2023-05-17 -> 2023-05-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


249/276 | S2 | 2023-05-25 -> 2023-06-02


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


250/276 | S2 | 2023-06-02 -> 2023-06-10


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


251/276 | S2 | 2023-06-10 -> 2023-06-18


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


252/276 | S2 | 2023-06-18 -> 2023-06-26


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


253/276 | S2 | 2023-06-26 -> 2023-07-04


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


254/276 | S2 | 2023-07-04 -> 2023-07-12


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


255/276 | S2 | 2023-07-12 -> 2023-07-20


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


256/276 | S2 | 2023-07-20 -> 2023-07-28


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


257/276 | S2 | 2023-07-28 -> 2023-08-05


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


258/276 | S2 | 2023-08-05 -> 2023-08-13


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


259/276 | S2 | 2023-08-13 -> 2023-08-21


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


260/276 | S2 | 2023-08-21 -> 2023-08-29


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


261/276 | S2 | 2023-08-29 -> 2023-09-06


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


262/276 | S2 | 2023-09-06 -> 2023-09-14


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


263/276 | S2 | 2023-09-14 -> 2023-09-22


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


264/276 | S2 | 2023-09-22 -> 2023-09-30


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


265/276 | S2 | 2023-09-30 -> 2023-10-08


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


266/276 | S2 | 2023-10-08 -> 2023-10-16


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


267/276 | S2 | 2023-10-16 -> 2023-10-24


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


268/276 | S2 | 2023-10-24 -> 2023-11-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


269/276 | S2 | 2023-11-01 -> 2023-11-09


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


270/276 | S2 | 2023-11-09 -> 2023-11-17


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


271/276 | S2 | 2023-11-17 -> 2023-11-25


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


272/276 | S2 | 2023-11-25 -> 2023-12-03


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


273/276 | S2 | 2023-12-03 -> 2023-12-11


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


274/276 | S2 | 2023-12-11 -> 2023-12-19


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


275/276 | S2 | 2023-12-19 -> 2023-12-27


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


276/276 | S2 | 2023-12-27 -> 2024-01-01


c:\Users\User\miniforge3\envs\et-fundacion\Lib\site-packages\ee\data.py:335: UserWarning: Your project has exceeded the compute quota of its noncommercial tier and is currently in restricted mode.
  warnings.warn(


Final station-period rows: 1380


## 9. Quality control and eligibility

Expected complete design:

- 138 MODIS periods
- 5 station footprints
- 2 optical sources
- 1,380 source × station × period rows

The historical diagnostic run produced 434 eligible HLS observations and 498 eligible Sentinel-2 observations at the 80% threshold. The counts below are recalculated, not imported.


In [31]:
expected_rows = (
    len(periods)
    * station_count
    * len(SOURCE_SCALES_M)
)

if len(observations) != expected_rows:
    raise ValueError(
        f"Incomplete calibration table: "
        f"{len(observations)} rows found, "
        f"{expected_rows} expected."
    )

observations["coverage_pct"] = pd.to_numeric(
    observations["coverage_pct"],
    errors="coerce",
)
observations["NDVI_p05"] = pd.to_numeric(
    observations["NDVI_p05"],
    errors="coerce",
)
observations["NDVI_p95"] = pd.to_numeric(
    observations["NDVI_p95"],
    errors="coerce",
)
observations["nonwater_pixel_count"] = pd.to_numeric(
    observations["nonwater_pixel_count"],
    errors="coerce",
)

# Convert explicit Earth Engine missing-value sentinels to NaN.
for column in ["NDVI_p05", "NDVI_p95"]:
    observations.loc[
        observations[column] <= -9990,
        column,
    ] = pd.NA

eligible = observations.loc[
    (observations["coverage_pct"] >= COVERAGE_THRESHOLD_PCT)
    & observations["NDVI_p05"].notna()
    & observations["NDVI_p95"].notna()
    & (observations["nonwater_pixel_count"] > 0)
].copy()

coverage_summary = (
    observations.groupby("source")
    .agg(
        total_observations=("source", "size"),
        mean_coverage_pct=("coverage_pct", "mean"),
        median_coverage_pct=("coverage_pct", "median"),
        eligible_observations=(
            "coverage_pct",
            lambda values: (
                values >= COVERAGE_THRESHOLD_PCT
            ).sum(),
        ),
    )
    .reset_index()
)

eligibility_summary = (
    eligible.groupby("source")
    .agg(
        eligible_observations=("source", "size"),
        stations=("station", "nunique"),
        ndvi_p05_min=("NDVI_p05", "min"),
        ndvi_p05_median=("NDVI_p05", "median"),
        ndvi_p05_max=("NDVI_p05", "max"),
        ndvi_p95_min=("NDVI_p95", "min"),
        ndvi_p95_median=("NDVI_p95", "median"),
        ndvi_p95_max=("NDVI_p95", "max"),
    )
    .reset_index()
)

display(coverage_summary.round(4))
display(eligibility_summary.round(6))


,source,total_observations,mean_coverage_pct,median_coverage_pct,eligible_observations
0,HLS,690,67.2431,100.0,434
1,S2,690,76.1477,100.0,498


,source,eligible_observations,stations,ndvi_p05_min,ndvi_p05_median,ndvi_p05_max,ndvi_p95_min,ndvi_p95_median,ndvi_p95_max
0,HLS,5,5,0.510267,0.661234,0.751651,0.827366,0.876207,0.908302


## 10. Calculate the global source-specific endmembers

Final calibration rule:

- `ndvi_low_endmember = P05(all eligible NDVI_p05)`
- `ndvi_high_endmember = P95(all eligible NDVI_p95)`


In [32]:
endmembers = (
    eligible.groupby("source")
    .agg(
        ndvi_low_endmember=(
            "NDVI_p05",
            lambda values: values.quantile(
                GLOBAL_LOW_QUANTILE
            ),
        ),
        ndvi_high_endmember=(
            "NDVI_p95",
            lambda values: values.quantile(
                GLOBAL_HIGH_QUANTILE
            ),
        ),
        n_observations=("source", "size"),
        n_stations=("station", "nunique"),
        period_start=("period_start", "min"),
        period_end=("period_end", "max"),
    )
    .reset_index()
)

endmembers["ndvi_range"] = (
    endmembers["ndvi_high_endmember"]
    - endmembers["ndvi_low_endmember"]
)

if set(endmembers["source"]) != {"HLS", "S2"}:
    raise ValueError(
        "Calibration must contain both HLS and S2."
    )

if not (endmembers["ndvi_range"] > 0).all():
    raise ValueError(
        "NDVI endmember range must be positive."
    )

if not (
    endmembers["ndvi_low_endmember"]
    .between(-1, 1)
    .all()
    and endmembers["ndvi_high_endmember"]
    .between(-1, 1)
    .all()
):
    raise ValueError(
        "NDVI endmembers must be within [-1, 1]."
    )

display(endmembers.round(6))


ValueError: Calibration must contain both HLS and S2.

## 11. Compare with the previously observed diagnostic values

These reference values are used **only as a reproducibility check**. They are not used to calculate or overwrite the new calibration.

Small differences can occur if repository preprocessing or Earth Engine source data change. Large differences should trigger methodological review before updating the production calibration.


In [ ]:
reference_values = pd.DataFrame(
    [
        {
            "source": "HLS",
            "reference_low": 0.416392,
            "reference_high": 0.908363,
        },
        {
            "source": "S2",
            "reference_low": 0.309061,
            "reference_high": 0.924045,
        },
    ]
)

reproducibility_check = endmembers.merge(
    reference_values,
    on="source",
    how="left",
)

reproducibility_check["low_difference"] = (
    reproducibility_check["ndvi_low_endmember"]
    - reproducibility_check["reference_low"]
)

reproducibility_check["high_difference"] = (
    reproducibility_check["ndvi_high_endmember"]
    - reproducibility_check["reference_high"]
)

display(
    reproducibility_check[
        [
            "source",
            "ndvi_low_endmember",
            "reference_low",
            "low_difference",
            "ndvi_high_endmember",
            "reference_high",
            "high_difference",
        ]
    ].round(6)
)


## 12. Save diagnostic outputs and the production calibration contract

The diagnostic CSV files remain under `outputs/` and may be ignored by Git.

The only file required by the production pipeline is:

`config/fvc_endmembers.json`

That JSON is generated from the values recalculated in this notebook.


In [ ]:
observations.to_csv(
    OBSERVATION_OUTPUT_PATH,
    index=False,
)

endmembers.to_csv(
    ENDMEMBER_TABLE_PATH,
    index=False,
)

source_config = {}

for row in endmembers.itertuples(index=False):
    source_config[row.source] = {
        "ndvi_low_endmember": float(
            row.ndvi_low_endmember
        ),
        "ndvi_high_endmember": float(
            row.ndvi_high_endmember
        ),
        "n_observations": int(
            row.n_observations
        ),
        "n_stations": int(
            row.n_stations
        ),
    }

calibration = {
    "calibration_name": (
        "fvc_global_ndvi_endmembers"
    ),
    "calibration_scope": (
        "external_diagnostic_prerequisite"
    ),
    "method": (
        "two_stage_global_percentile"
    ),
    "coverage_threshold_pct": (
        COVERAGE_THRESHOLD_PCT
    ),
    "coverage_bands": [
        "Green",
        "Red",
        "NIR",
    ],
    "water_exclusion": {
        "index": "NDWI",
        "rule": "exclude NDWI > 0",
    },
    "within_footprint_period": {
        "low_candidate_percentile": (
            WITHIN_PERIOD_LOW_PERCENTILE
        ),
        "high_candidate_percentile": (
            WITHIN_PERIOD_HIGH_PERCENTILE
        ),
    },
    "across_all_valid_observations": {
        "low_quantile": GLOBAL_LOW_QUANTILE,
        "high_quantile": GLOBAL_HIGH_QUANTILE,
        "temporal_scheme": "global",
        "sources_calibrated_separately": True,
    },
    "calibration_period": {
        "start": min(period_starts).isoformat(),
        "end": max(
            period_end
            for _, period_end in periods
        ).isoformat(),
    },
    "sources": source_config,
}

with PIPELINE_CONFIG_PATH.open(
    "w",
    encoding="utf-8",
) as file:
    json.dump(
        calibration,
        file,
        indent=2,
        ensure_ascii=False,
    )
    file.write("\n")

print(
    "Saved observations:",
    OBSERVATION_OUTPUT_PATH,
)
print(
    "Saved endmember table:",
    ENDMEMBER_TABLE_PATH,
)
print(
    "Saved pipeline calibration:",
    PIPELINE_CONFIG_PATH,
)

print("\nFinal calibration:")
for source, values in calibration["sources"].items():
    print(
        f"{source}: "
        f"low={values['ndvi_low_endmember']:.6f}, "
        f"high={values['ndvi_high_endmember']:.6f}, "
        f"n={values['n_observations']}"
    )


## Production pipeline requirement

This notebook is **not** part of the operational ET/FVC pipeline.

The production pipeline must treat `config/fvc_endmembers.json` as a prerequisite and should:

1. fail if the file does not exist;
2. fail if the requested optical source is absent;
3. validate `-1 <= low < high <= 1`;
4. use the stored source-specific endmembers;
5. never recalculate endmembers silently during production execution.

If the calibration methodology changes, this notebook should be rerun explicitly and the updated JSON reviewed and versioned.
